Add the authentication service

IBruteDude opened the pull request.

IBruteDude committed:
Added ErrorHandling types

IBruteDude committed:
Added and implemented the AuthenticationService

ahmedeldamaty20 reviewed (changes requested):
Make pull request based on feature/authentication not main branch

IBruteDude committed:
Fixes:
- Relocated the Auth service implementation
- Removed the Update method
- Implemented logout method
- Corrected the incomplete app user response

IBruteDude committed:
Changes:
- Refactored the Status class
- Added the more descriptive StatusCode enum
- Refactored AuthenticationService to use StatusCode constructor

IBruteDude committed:
Changes:
- Refactored the Result pattern
- Refactored the AuthenticationService to use the new Result class constructs
